x/upgrade/keeper.go

```diff
@@ -146,95 +120,38 @@ func (keeper *keeper) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock)
+				ctx.Logger().Error(fmt.Sprintf("UPGRADE \"%s\" NEEDED at height %d: %s", plan.Name, blockHeight, plan.Info))
+				panic("UPGRADE REQUIRED!")
 			}
-			panic("UPGRADE REQUIRED!")
 		}
 	}
-	// TODO will upgraded is disabled for now because of inconsistent behavior
-	//else if justRetrievedFromCache {
-	//	// In this case we are notifying the system that an upgrade is planned but not scheduled to happen yet
-	//	if keeper.willUpgrader != nil {
-	//		keeper.willUpgrader(ctx, keeper.plan)
-	//	} else {
-	//		DefaultWillUpgrader(ctx, keeper.plan)
-	//	}
-	//}
-}
-
 
-// DefaultWillUpgrader asynchronously runs a script called prepare-upgrade from $COSMOS_HOME/config if such a script exists,
-// with plan serialized to JSON as the first argument and the current block height as the second argument.
-// The environment variable $COSMOS_HOME will be set to the home directory of the daemon.
-func DefaultWillUpgrader(ctx sdk.Context, plan Plan) {
-	CallUpgradeScript(ctx, plan, "prepare-upgrade", true)
-}
-
-// DefaultOnUpgrader synchronously runs a script called do-upgrade from $COSMOS_HOME/config if such a script exists,
-// with plan serialized to JSON as the first argument and the current block height as the second argument.
-// The environment variable $COSMOS_HOME will be set to the home directory of the daemon.
-func DefaultOnUpgrader(ctx sdk.Context, plan Plan) {
-	CallUpgradeScript(ctx, plan, "do-upgrade", false)
-}
-
-// CallUpgradeScript runs a script called script from $COSMOS_HOME/config if such a script exists,
-// with plan serialized to JSON as the first argument and the current block height as the second argument.
-// The environment variable $COSMOS_HOME will be set to the home directory of the daemon.
-// If async is true, the command will be run in a separate go-routine.
-func CallUpgradeScript(ctx sdk.Context, plan Plan, script string, async bool) {
-	f := func() {
-		home := viper.GetString(cli.HomeFlag)
-		file := filepath.Join(home, "config", script)
-		ctx.Logger().Info(fmt.Sprintf("Looking for upgrade script %s", file))
-		if _, err := os.Stat(file); err == nil {
-			ctx.Logger().Info(fmt.Sprintf("Applying upgrade script %s", file))
-			err = os.Setenv("COSMOS_HOME", home)
-			if err != nil {
-				ctx.Logger().Error(fmt.Sprintf("Error setting env var COSMOS_HOME: %v", err))
-			}
-
-
-			planJson, err := json.Marshal(plan)
-			if err != nil {
-				ctx.Logger().Error(fmt.Sprintf("Error marshaling upgrade plan to JSON: %v", err))
-			}
-			cmd := exec.Command(file, string(planJson), fmt.Sprintf("%d", ctx.BlockHeight()))
-			cmd.Stdout = logWriter{ctx, script, false}
-			cmd.Stderr = logWriter{ctx, script, false}
-			err = cmd.Run()
-			if err != nil {
-				ctx.Logger().Error(fmt.Sprintf("Error running script %s: %v", file, err))
-			}
+	// in every case (after upgrade, or no upgrade), ensure we are not too recent version
+	//all registered handlers should be in done
```

Comment:
I don't think it is necessary to make sure all registered handlers are in done. The reason for this is that there may be some registered handlers which were not applied directly to a given zone but are still in that zone's binary. Say a zone is starting up some point after genesis.

I think the sufficient check is - given the current plan that scheduled, if there is a handler registered for that plan and it is not yet time to upgrade, then panic.

Reply:
Okay, I will make it less strict